### PR TITLE
Add Rust provenance module

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Currently, it's very easy to pass the wrong `AccountId` to methods. This PR solves this by wrapping `AccountId` in newtype wrapper.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1203193667254522/f) _(internal)_
- https://github.com/hashintel/hash/pull/1197#discussion_r999207507

## 🔍 What does this change?

- Add newtype wrapper for `OwnedById`, `CreatedById`, `UpdatedById`, and `RemovedById`
- Move `AccountId` to `identifier` module
- ~~Move provenance-related structs into the provenance module (`Persisted*`)~~

## ⚠️  Known issues

~~Even though I only added very few things (`*ById`s) I moved some code around and GitHub was not able to detect those moves. For reviewing I recommend going through the commit list~~

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/provenance/ids.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use serde::{Deserialize, Serialize};
+use utoipa::{openapi::Schema, ToSchema};
+use uuid::Uuid;
+
+#[derive(
+    Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, ToSchema, FromSql, ToSql,
+)]
+#[repr(transparent)]
+#[postgres(transparent)]
+pub struct AccountId(Uuid);
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/provenance/ids.rs:13]: I don't think this should live in `provenance` this should probably live inside `identifier.rs`. The others are records for provenance and auditing, but this is a lot more foundational and global.

File: packages/graph/hash_graph/lib/graph/src/provenance/ontology.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    #[must_use]
+    pub const fn owned_by_id(&self) -> OwnedById {
+        self.owned_by_id
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct PersistedOntologyMetadata {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/provenance/ontology.rs:38]: I don't think these should have moved to provenance, let's talk about this offline.

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
Thanks! This must have been tedious 🙃 

The new-types are exactly what I wanted, I don't think we should have moved the metadata and identifier structs to the provenance modules though. I think provenance records are more this global definition of the kind of metadata we're storing for operations (e.g. "Bob did X") but the actual metadata structs, and especially the identifiers, are things that belong to the records and _use/contain_ the provenance fields.

I think we should probably revert the last few commits

File: packages/graph/hash_graph/lib/graph/src/provenance/ontology.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/provenance/ontology.rs]: This is why I made separate commits for this. I reverted these.

**Comment by TimDiekmann** [comment]
Thanks, Alfie, I applied the review.

**Review by Alfred-Mountfield — COMMENTED** [review]
Thanks Tim, it looks good! Do we perhaps want to move `provenance` into `shared`?

**Comment by TimDiekmann** [comment]
Good idea, applied in aa38b06b601a4ee7d63126baafe826efc7221e27

**Review by Alfred-Mountfield — APPROVED** [review]
🚀

**Pull request merged**